chore: modernize project structure, build system, and tooling (src layout + uv + ruff + CI)

tavallaie opened the pull request:
This PR brings the repository up to modern 2025 Python packaging and development standards:

### Major changes
- Migrated to `src/` layout 
- Fully switched to `pyproject.toml` + `uv` for dependency management and builds (removed `setup.py`, `setup.cfg`, `requirements.txt`)
- Added comprehensive `Makefile` with `format`, `lint`, `test`, and `check` targets
- Introduced Ruff for formatting + linting, added dev dependency group, and applied fixes across the entire codebase
- Added `.pre-commit-config.yaml` and GitHub Actions workflow for linting, testing, and publishing
- Added `.gitignore` and cleaned up `__pycache__/` directories

### Publishing improvements
The new GitHub Actions workflow is ready for Trusted Publishing on PyPI (no more manual API tokens).
After this PR is merged, a maintainer can enable it in a few seconds here:
https://docs.pypi.org/trusted-publishers/adding-a-publisher/

### Development workflow (uv-based)
#### install dependencies + dev extras
`uv sync --all-groups --all-extras `

#### for formatting and fixing:

`make format `

or

#### only linting the codebase

`make lint `

#### run the test suite

`make test`                

**After adding the Trusted Publisher, bump the version in `pyproject.toml`  to publish to `PyPI`.**

tavallaie committed:
add .gitignore and remove __pycache__/

tavallaie committed:
feat: migrate to src layout and rename package to resemble_perth

- Move all Python modules from `perth/` to `src/resemble_perth/`
- Align package name with PyPI distribution name (`resemble-perth`)
- Adopt standard `src/` layout for better isolation and packaging hygiene
- Add `.python-version` for Python version consistency

This change supports modern Python packaging practices and prepares the project
for reliable builds with `uv`, `pip`, and other PEP 517-compliant tools.

tavallaie committed:
chore: migrate to pyproject.toml and adopt uv for dependency management

- Replace setup.py, setup.cfg, and requirements.txt with standardized pyproject.toml
- Declare project metadata, dependencies, and build configuration per PEP 621
- Introduce uv.lock for reproducible, fast dependency resolution using uv
- Fully transition to modern Python packaging tooling

tavallaie committed:
chore: rename src/resemble_perth → src/perth for correct import name

tavallaie committed:
feat: add Makefile for formatting, linting, and testing

tavallaie committed:
chore: add dev dependency group for pre-commit and ruff

tavallaie committed:
chore: apply Ruff formatting and fix linting issues across codebase

- Fixed E722: replaced bare `except` with `except AttributeError`
- Fixed E721: used `is` for type comparisons (e.g., `is bool`)
- Applied consistent code formatting with Ruff
- Updated all core modules in `src/perth/`

tavallaie committed:
adding pre-commit config

tavallaie committed:
feat: add GitHub Actions workflow for linting, testing, and publishing

tavallaie commented:
Also, I think this PR will solve **#7 Missing dependency**

tavallaie commented:
@johnmeade, can you review this PR? This PR helps to update the library faster and also has a better dependency management.

manmay-nakhashi commented:
@tavallaie looking at it

manmay-nakhashi commented:
lot's of file changes can you remove all the __pychache__ and everything cleanup the branch and i'll have a look again

tavallaie commented:
I already removed all the __pycache__ 
Put that directory in gitignore
Also I implemented src/ structure for better maintaining with github action so you see a lot of files changed. 
Also I updated dependencies. 
Another change adding pre-commit to implement linting.

tavallaie commented:
@manmay-nakhashi @johnmeade, do you need any comments from me on code, or more explanation here about the changes I made? Maybe it will make it easier to review.

johnmeade commented:
Wow this is quite extensive, but thanks for cleaning it up. I'll look into this soon

johnmeade commented on pyproject.toml:
```diff
@@ -0,0 +1,78 @@
+requires-python = ">=3.10"
+authors = [
+    { name = "Resemble AI", email = "team@resemble.ai" },
+    { name = "Adityu", email = "aditya@resemble.ai" },
```
```suggestion
    { name = "Aditya", email = "aditya@resemble.ai" },
    { name = "John", email = "john@resemble.ai" },
    { name = "Manmay", email = "manmay@resemble.ai" },
```

johnmeade commented on .gitignore:
```diff
@@ -0,0 +1,379 @@
+# SageMath parsed files
+*.sage.py
+
+# Environments
```
extremely minor comment, there is a lot of duplication in here

tavallaie replied:
I copy that from a template, I'll fix them tonight

tavallaie committed:
Update pyproject.toml

Co-authored-by: John Meade <26562807+johnmeade@users.noreply.github.com>

tavallaie committed:
chore: triming .gitignore

tavallaie commented:
@johnmeade I think, it is ready for the code review.

johnmeade reviewed (approved):
Awesome, thanks for everything, I gave it a test locally and it works fine

tavallaie commented:
@johnmeade, Thanks for merging my `PR`
now if you follow this link:
https://docs.pypi.org/trusted-publishers/adding-a-publisher/
you can update Perth library with github action on `PyPI`

johnmeade commented:
Neat! We may improve the model in the future depending on demand, so perhaps we will try that out later
